### PR TITLE
Implement Espirit, make some transforms nn.Modules, add new transforms

### PR DESCRIPTION
**Pull request by georgeyiasemis**
* Implements Espirit calibration sensitivity map estimation and methods to calculate maximum eigenvalue numerically (e.g. powermethod)
* Some transforms are now nn.Modules and can be implemented batch-wise
* MRI transforms split to `pre-transforms` and `post-transforms`. Plan is to call `pre-transforms` before collation and `post-transforms` after collation on device [not implemented yet, functionality now is as it was].
* Adding augmentation transforms: `RandomFlip` and `RandomRotation`:
    * They work by backprojecting the k-space to the image domain, transforming it, and then projecting back to k-space domain.

### CODE REVIEW

File: direct/algorithms/mri.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
+# coding=utf-8
```
Comment (jonasteuwen) [inline direct/algorithms/mri.py:1]: Can you give this folder a more descriptive title? What does "mri.py" do?

File: direct/algorithms/__init__.py [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# coding=utf-8
+# Copyright (c) DIRECT Contributors
```
Comment (jonasteuwen) [inline direct/algorithms/__init__.py:2]: Pylint likely requires you to give a module description. Can you try to run `pylint` on this submodule?

File: direct/algorithms/mri.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+# Copyright (c) DIRECT Contributors
+
+from typing import Any, Callable, Dict
+
```
Comment (jonasteuwen) [inline direct/algorithms/mri.py:5]: Module description

File: direct/algorithms/mri.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    def __init__(
+        self,
+        backward_operator: Callable,
+        threshold: float = 0.05,
+        kernel_size: int = 6,
+        crop: float = 0.95,
+        max_iter: int = 100,
+        kspace_key: str = "masked_kspace",
+    ):
+        """Inits :class:`EstimateSensitivityMap`.
```
Comment (jonasteuwen) [inline direct/algorithms/mri.py:39]: ? isn't this a different class

File: direct/algorithms/mri.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        Returns
+        -------
+        torch.Tensor
+            Cropped k-space.
+        """
+        nonzero_idxs = torch.nonzero(acs_mask)
+        x, y = nonzero_idxs[..., 0], nonzero_idxs[..., 1]
+        xl, xr = x.min(), x.max()
+        yl, yr = y.min(), y.max()
+        return kspace[:, xl : xr + 1, yl : yr + 1]
```
Comment (jonasteuwen) [inline direct/algorithms/mri.py:84]: Don't we have this somewhere already? If so, wouldn't it make sense to have this as a separate function?

File: direct/algorithms/optimization.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
+# coding=utf-8
+# Copyright (c) DIRECT Contributors
+
```
Comment (jonasteuwen) [inline direct/algorithms/optimization.py:3]: Module docstring

File: direct/algorithms/mri.py (hunk 4 above)
Comment (jonasteuwen) [reply direct/algorithms/mri.py:39]: Can't kspace_key be an Enum, just like https://github.com/kornia/kornia/blob/b4ab100a1d2b87f8ae290dbd227c407c06f1d3d3/kornia/constants.py#L125

File: direct/algorithms/optimization.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    Solves for:
+    .. math:: A x = b
+    where A is a Hermitian linear operator.
+    Args:
+        A (Linop or function): Linop or function to compute A.
+        b (array): Observation.
+        x (array): Variable.
+        max_iter (int): Maximum number of iterations.
+        tol (float): Tolerance for stopping condition.
+    """
```
Comment (jonasteuwen) [inline direct/algorithms/optimization.py:95]: This is not the way we do docstrings?

File: direct/data/datasets_config.py [hunk 8]
```diff
@@ -19,8 +19,19 @@ class TransformsConfig(BaseConfig):
...
+    random_rotation_probability: Optional[float] = 0.5
+    random_flip: bool = False
+    random_flip_type: Optional[str] = "random"
+    random_flip_probability: Optional[float] = 0.5
+    estimate_sensitivity_maps: bool = True
+    sensitivity_maps_type: str = "rss_estimate"
+    sensitivity_maps_espirit_threshold: Optional[float] = 0.05
+    sensitivity_maps_espirit_kernel_size: Optional[int] = 6
+    sensitivity_maps_espirit_crop: Optional[float] = 0.95
+    sensitivity_maps_espirit_max_iters: Optional[int] = 30
```
Comment (jonasteuwen) [inline direct/data/datasets_config.py:34]: Not a fan. Can't we have transforms a different config?| Like `transforms: TransformConfig`

File: direct/utils/__init__.py [hunk 9]
```diff
@@ -338,7 +338,7 @@ def return_func(*args, **kwargs):
     return return_func
 
 
-class DirectTransform:
+class DirectTransform(torch.nn.Module):
     """Direct transform class.
 
     Defines :meth:`__repr__` method for Direct transforms.
```
Comment (jonasteuwen) [inline direct/utils/__init__.py:344]: DirectModule makes more sense as a name

**Review by jonasteuwen — COMMENTED** [review]
Some initial comments

File: direct/algorithms/mri.py (hunk 5 above)
Comment (georgeyiasemis) [reply direct/algorithms/mri.py]: Done

File: direct/algorithms/mri.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+    def calculate_sensitivity_map(self, acs_mask: torch.Tensor, kspace: torch.Tensor) -> torch.Tensor:
+        # pylint: disable=too-many-locals
+        ndim = kspace.ndim - 2
+        spatial_size = kspace.shape[1:-1]
+
+        # Used in case the k-space is padded (e.g. for batches)
```
[the remark this answers is not in the document]
Comment (georgeyiasemis) [reply direct/algorithms/mri.py]: Done

File: direct/algorithms/optimization.py (hunk 6 above)
Comment (georgeyiasemis) [reply direct/algorithms/optimization.py:3]: Done

File: direct/algorithms/optimization.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+import torch
+from torch import nn
+
+
+class Algorithm(nn.Module):
+    def __init__(self, max_iter: int = 30):
+        self.max_iter = max_iter
+        self.iter = 0
+        super().__init__()
+
```
[the remark this answers is not in the document]
Comment (georgeyiasemis) [reply direct/algorithms/optimization.py]: Done

File: direct/algorithms/optimization.py (hunk 7 above)
Comment (georgeyiasemis) [reply direct/algorithms/optimization.py]: Done

File: direct/algorithms/optimization.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        A (Linop or function): Linop or function to compute A.
+        b (array): Observation.
+        x (array): Variable.
+        max_iter (int): Maximum number of iterations.
+        tol (float): Tolerance for stopping condition.
+    """
+
+    def __init__(
+        self,
+        A: Callable,
```
[the remark this answers is not in the document]
Comment (georgeyiasemis) [reply direct/algorithms/optimization.py]: Method deleted for now

File: direct/utils/__init__.py (hunk 9 above)
Comment (georgeyiasemis) [reply direct/utils/__init__.py]: True

Comment (georgeyiasemis) [reply direct/utils/__init__.py]: Done

File: direct/algorithms/mri.py (hunk 4 above)
Comment (georgeyiasemis) [reply direct/algorithms/mri.py:41]: Done

File: direct/data/datasets_config.py (hunk 8 above)
Comment (georgeyiasemis) [reply direct/data/datasets_config.py:34]: Will do in future PR

**Pull request closed without merging**